module/zfs/dsl_scan.c

```diff
@@ -236,24 +236,40 @@ typedef enum {
  */
 typedef struct scan_io {
 	/* fields from blkptr_t */
-	uint64_t		sio_offset;
 	uint64_t		sio_blk_prop;
 	uint64_t		sio_phys_birth;
 	uint64_t		sio_birth;
 	zio_cksum_t		sio_cksum;
-	uint32_t		sio_asize;
+	uint32_t		sio_nr_dvas;
 
 	/* fields from zio_t */
-	int			sio_flags;
+	uint32_t		sio_flags;
 	zbookmark_phys_t	sio_zb;
 
 	/* members for queue sorting */
 	union {
-		avl_node_t	sio_addr_node; /* link into issueing queue */
+		avl_node_t	sio_addr_node; /* link into issuing queue */
 		list_node_t	sio_list_node; /* link for issuing to disk */
 	} sio_nodes;
+
+	/*
+	 * There may be up to SPA_DVAS_PER_BP DVAs here from the bp,
+	 * depending on how many were in the original bp. This array
+	 * must go at the end of the struct to allow this for the
+	 * variable size.
+	 */
+	dva_t			sio_dva[0];
 } scan_io_t;
 
+#define	SIO_SET_OFFSET(sio, x)		DVA_SET_OFFSET(&(sio)->sio_dva[0], x)
+#define	SIO_SET_ASIZE(sio, x)		DVA_SET_ASIZE(&(sio)->sio_dva[0], x)
+#define	SIO_GET_OFFSET(sio)		DVA_GET_OFFSET(&(sio)->sio_dva[0])
+#define	SIO_GET_ASIZE(sio)		DVA_GET_ASIZE(&(sio)->sio_dva[0])
```

Comment:
This is a little confusing since BP_GET_ASIZE() is the sum of the dva's asizes.  I think that this code might still be correct, since you're really only scrubbing dva[0], and for the most part the others are just along for the ride.  Separate scan_io_t's will represent the other DVA's.  The best explanation of this that I could find was in bp2sio().  It might be helpful to explain this in more detail here (maybe where sio_dva[] is declared, and then also mention here that the ASIZE is only of the first dva because we are only scrubbing it?)

I think it could be arguably clearer to express this by having scan_io_t contain 
```
dva_t sio_target_dva; 
dva_t sio_additional_dvas[2];
```
rather than having dva[0] be implicitly special.

Reply:
What if i change the comment to this:
```
	/*
	 * There may be up to SPA_DVAS_PER_BP DVAs here from the bp,
	 * depending on how many were in the original bp. Only the
	 * first DVA is really used for sorting and issuing purposes.
	 * The other DVAs (if provided) simply exist so that the zio
	 * layer can find additional copies to repair from in the
	 * event of an error. This array must go at the end of the
	 * struct to allow this for the variable number of elements.
	 */
```

Reply:
The macro's below the struct definition are the only bits of code that actually refer to sio_dva[0] so i don't think there's a risk of causing confusion.

Reply:
OK